src/config/cmd.rs

```diff
@@ -166,4 +226,159 @@ mod tests {
+        sample::subsequence(SourceDescription::types(), 2..=4)
+    }
+
+    /// Select any 2-4 transforms
+    fn arb_transforms() -> impl Strategy<Value = Vec<&'static str>> {
+        sample::subsequence(TransformDescription::types(), 2..=4)
+    }
+
+    /// Select any 2-4 sinks
+    fn arb_sinks() -> impl Strategy<Value = Vec<&'static str>> {
+        sample::subsequence(SinkDescription::types(), 2..=4)
+    }
+
+    fn create_config_source(sources: &[&str], transforms: &[&str], sinks: &[&str]) -> String {
+        // This creates a string in the syntax expected by the `vector generate`
+        // command whose internal mechanics we are using to create valid Vector
+        // configurations.
+        //
+        // Importantly, we have to name the components (in this case, simply by
+        // their type as each type of component is guaranteed to only appear
+        // once), because (in some tests) we'd like to shuffle the configuration
+        // later in a way that does not change its actual semantics. Otherwise,
+        // an autogenerated ID like `source0` could correspond to different
+        // sources depending on the ordering of the `vector generate` input.
+        //
+        // We also append a fixed `remap` transform to the transforms list. This
+        // ensures sink inputs are consistent since `generate` uses the last
+        // transform the input for each sink.
+        let generate_config_str = format!(
+            "{}/{}/{}",
+            sources
+                .iter()
+                .map(|source| format!("{}:{}", source, source))
+                .collect::<Vec<_>>()
+                .join(","),
+            transforms
+                .iter()
+                .map(|transform| format!("{}:{}", transform, transform))
+                .chain(vec!["manually-added-remap:remap".to_string()])
+                .collect::<Vec<_>>()
+                .join(","),
+            sinks
+                .iter()
+                .map(|sink| format!("{}:{}", sink, sink))
+                .collect::<Vec<_>>()
+                .join(","),
+        );
+        generate_example(
+            false,
+            generate_config_str.as_ref(),
+            &None,
+            TransformInputsStrategy::All,
+        )
+        .expect("invalid config generated")
+    }
+
+    proptest! {
+        #[test]
+        /// Output should be the same regardless of input config ordering
+        fn output_has_consistent_ordering(mut sources in arb_sources(), mut transforms in arb_transforms(), mut sinks in arb_sinks()) {
+            let config_source = create_config_source(sources.as_ref(), transforms.as_ref(), sinks.as_ref());
+
+            // Shuffle the ordering of components which shuffles the order in
+            // which items appear in the final config
+            let mut rng = thread_rng();
+            sources.shuffle(&mut rng);
+            transforms.shuffle(&mut rng);
+            sinks.shuffle(&mut rng);
```

Comment:
One of the important things that a property test does is 'shrink' failing inputs but for shrinking to work the property test machinery must be able to re-run tests deterministically to confirm failure. You should remove this source of randomness and rely on the generator -- your `arb_sources` etc -- to handle shuffling.

Reply:
Gotcha, great catch. For this test case it's important to keep the same components (in different orderings) for each individual run, so I was using the generators to create just the initial set of random components. It occurs to me that swapping this for a deterministic shuffle could also work given we have a random input -- does that make sense?

Reply:
> For this test case it's important to keep the same components (in different orderings) for each individual run,

No, I'm not totally sure what you mean here. 

Perhaps I'm misunderstanding how your generators work, but they should be the ones that 'shuffle' the sources, transforms etc. You're generating randomly -- I think -- ordered components and then non-deterministically shuffling them with thread randomness.

Reply:
Ah sorry, to be clearer, I'm suggesting
- Use the generators to create the initial set of randomly ordered components (as it currently is)
- Deterministically shuffle (i.e. reverse the array) the components to produce a different ordering (rather than the random shuffle)

The test is then asserting that, after formatting these two equivalent but differently ordered configurations, the ordering is irrelevant.

I may have misinterpreted your original generator suggestion. I was trying to clarify that if I used the generators to produce two sets of components, the components would then be different.

Reply:
Ah ha, I missed the purpose of that `config_source` on line 340. 

Yes, alright. You need to drop the `thread_rng` but you have two options as I see it, since shuffling in-test is an important feature here. You could fiddle with your generators so that it returns the components twice, once in the 'unshuffled' form and the other in 'shuffled' form, likely using https://docs.rs/proptest/1.0.0/proptest/strategy/struct.Shuffle.html _or_ you could pass in a seed for your random generator. Both approaches work, both are deterministic and allow for shrinking.

Reply:
Switched to a seeded random generator in [5b336cf](https://github.com/vectordotdev/vector/pull/12713/commits/5b336cfaf73b146edde7f83611212e124cc7d4f4)